extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/ReadMongoP.java

```diff
@@ -0,0 +1,482 @@
+        private final List<Bson> aggregates;
+        private Traverser<Document> delegate;
+        private ObjectId lastKey;
+
+        private BatchMongoReader(
+                String databaseName,
+                String collectionName,
+                FunctionEx<Document, I> mapItemFn,
+                List<Bson> aggregates) {
+            super(databaseName, collectionName);
+            this.mapItemFn = mapItemFn;
+            this.aggregates = aggregates;
+        }
+
+        @Override
+        void onConnect(MongoClient mongoClient, boolean supportsSnapshots) {
+            List<Bson> aggregateList = new ArrayList<>(aggregates);
+            if (supportsSnapshots && !hasSorts(aggregateList)) {
+                aggregateList.add(sort(ascending("_id")).toBsonDocument());
+            }
+            if (totalParallelism > 1) {
+                aggregateList.addAll(0, partitionAggregate(totalParallelism, processorIndex, false));
+            }
+            if (collection != null) {
+                this.delegate = delegateForCollection(collection, aggregateList);
+            } else if (database != null) {
+                this.delegate = delegateForDb(database, aggregateList);
+            } else {
+                final MongoClient clientLocal = mongoClient;
+                this.delegate = traverseIterable(mongoClient.listDatabaseNames())
+                        .flatMap(name -> {
+                            MongoDatabase db = clientLocal.getDatabase(name);
+                            return delegateForDb(db, aggregateList);
+                        });
+            }
+            checkNotNull(this.delegate, "unable to construct Mongo traverser");
+        }
+
+        private boolean hasSorts(List<Bson> aggregateList) {
+            return aggregateList.stream().anyMatch(agg -> agg.toBsonDocument().get("$sort") != null);
+        }
+
+        private Traverser<Document> delegateForCollection(MongoCollection<Document> collection, List<Bson> aggregateList) {
+            return traverseIterable(collection.aggregate(aggregateList));
+        }
+
+        private Traverser<Document> delegateForDb(MongoDatabase database, List<Bson> aggregateList) {
+            MongoIterable<String> collectionsIterable = database.listCollectionNames();
+
+            return traverseIterable(collectionsIterable)
+                    .flatMap(colName -> delegateForCollection(database.getCollection(colName), aggregateList));
+        }
+
+        @Nonnull
+        @Override
+        public Traverser<I> nextChunkTraverser() {
+            List<I> chunk = new ArrayList<>(BATCH_SIZE);
+            Document doc;
+            Document lastItem = null;
+            while ((doc = delegate.next()) != null) {
+                chunk.add(mapItemFn.apply(doc));
+                lastItem = doc;
+            }
+            lastKey = lastItem == null ? null : lastItem.getObjectId("__id");
+
+            return Traversers.traverseIterable(chunk);
```

Comment:
Why batch these items and not return the Traverser you already have (`delegate`) - there is com.hazelcast.jet.Traverser#map to remap the item.

Reply:
6f18195af4f9493afd5598b62b421744331792b8